### PR TITLE
feat: Adds support for client side prerequisite events.

### PR DESCRIPTION
**Pull request by tanderson-ld**
**Requirements**

- [x] I have added test coverage for new or changed functionality
- [x] I have followed the repository's [pull request submission guidelines](../blob/main/CONTRIBUTING.md#submitting-pull-requests)
- [x] I have validated my changes against all supported platform versions

**Related issues**

SDK-687

### CODE REVIEW

File: lib/sdk/server/src/main/java/com/launchdarkly/sdk/server/Evaluator.java [hunk 1]
```diff
@@ -145,22 +146,40 @@ EvalResult evaluate(FeatureFlag flag, LDContext context, @Nonnull EvaluationReco
...
 
       if (state.bigSegmentsStatus != null) {
-        return result.withReason(
+        result = result.withReason(
             result.getReason().withBigSegmentsStatus(state.bigSegmentsStatus)
         );
       }
+
+      // TODO: these changes have reduced throughput, can we optimize this a bit.  Perhaps by calling constructor
+      // with all parameters instead of using multiple calls in this immutable style
```
Comment (tanderson-ld) [inline lib/sdk/server/src/main/java/com/launchdarkly/sdk/server/Evaluator.java:162]: Do TODO

File: lib/sdk/server/src/test/java/com/launchdarkly/sdk/server/LDClientEvaluationTest.java [hunk 2]
```diff
@@ -653,4 +655,45 @@ public void allFlagsStateReturnsEmptyStateIfClientAndStoreAreNotInitialized() th
       assertFalse(state.isValid());
     }
   }
+
+  @Test
```
Comment (tanderson-ld) [inline lib/sdk/server/src/test/java/com/launchdarkly/sdk/server/LDClientEvaluationTest.java:659]: Need to add more test cases.

**Review by louis-launchdarkly — COMMENTED** [review]
So far this looks good, let me know when the TODO are done, I can approve.

**Comment by tanderson-ld** [comment]
TODO: performance bench testing to validate prerequisite list allocation performance.

**Comment by tanderson-ld** [comment]
> TODO: performance bench testing to validate prerequisite list allocation performance.

Performance isn't significantly impacted

Prereq branch
  Benchmark                                                                   Mode  Cnt     Score    Error   Units
  LDClientEvaluationBenchmarks.allFlagState                                   avgt  100   407.517 ±  5.443   ns/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.alloc.rate                    avgt  100  2489.288 ± 31.002  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.alloc.rate.norm               avgt  100  1594.846 ±  0.004    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Eden_Space           avgt  100  2371.667 ± 33.377  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Eden_Space.norm      avgt  100  1519.796 ± 14.185    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Survivor_Space       avgt  100     0.013 ±  0.001  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Survivor_Space.norm  avgt  100     0.008 ±  0.001    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.count                         avgt  100  1911.000           counts
  LDClientEvaluationBenchmarks.allFlagState:·gc.time                          avgt  100  1612.000               ms

  Benchmark                                                                   Mode  Cnt     Score    Error   Units
  LDClientEvaluationBenchmarks.allFlagState                                   avgt  100   396.184 ±  3.652   ns/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.alloc.rate                    avgt  100  2558.533 ± 22.781  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.alloc.rate.norm               avgt  100  1594.840 ±  0.004    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Eden_Space           avgt  100  2441.965 ± 34.891  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Eden_Space.norm      avgt  100  1521.985 ± 14.795    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Survivor_Space       avgt  100     0.014 ±  0.001  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Survivor_Space.norm  avgt  100     0.008 ±  0.001    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.count                         avgt  100  1898.000           counts
  LDClientEvaluationBenchmarks.allFlagState:·gc.time                          avgt  100  1659.000               ms

Main
  Benchmark                                                                   Mode  Cnt     Score    Error   Units
  LDClientEvaluationBenchmarks.allFlagState                                   avgt  100   407.267 ±  3.380   ns/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.alloc.rate                    avgt  100  2517.584 ± 19.629  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.alloc.rate.norm               avgt  100  1613.513 ±  0.003    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Eden_Space           avgt  100  2394.218 ± 30.173  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Eden_Space.norm      avgt  100  1534.405 ± 14.654    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Survivor_Space       avgt  100     0.014 ±  0.001  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Survivor_Space.norm  avgt  100     0.009 ±  0.001    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.count                         avgt  100  1910.000           counts
  LDClientEvaluationBenchmarks.allFlagState:·gc.time                          avgt  100  1629.000               ms

  Benchmark                                                                   Mode  Cnt     Score    Error   Units
  LDClientEvaluationBenchmarks.allFlagState                                   avgt  100   391.854 ±  4.775   ns/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.alloc.rate                    avgt  100  2588.506 ± 29.915  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.alloc.rate.norm               avgt  100  1594.841 ±  0.003    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Eden_Space           avgt  100  2509.273 ± 39.834  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Eden_Space.norm      avgt  100  1545.881 ± 15.164    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Survivor_Space       avgt  100     0.014 ±  0.001  MB/sec
  LDClientEvaluationBenchmarks.allFlagState:·gc.churn.G1_Survivor_Space.norm  avgt  100     0.009 ±  0.001    B/op
  LDClientEvaluationBenchmarks.allFlagState:·gc.count                         avgt  100  1930.000           counts
  LDClientEvaluationBenchmarks.allFlagState:·gc.time                          avgt  100  1664.000               ms

**Pull request merged**